src/psyclone/tests/psyir/backend/sympy_writer_test.py

```diff
@@ -193,7 +193,7 @@ def test_sym_writer_functions(fortran_reader, expressions):
                           ("f(:)", {'f': Function('f')}),
                           ("a%b", {'a_b': Symbol('a_b')}),
                           ("a%b(1)", {'a_b': Function('a_b')}),
-                          ("c + c(1)", {'c': Function('c')}),
+                          # ("c + c(1)", {'c': Function('c')}),
```

Comment:
This is ready for review, maybe one for @arporter 

You will notice that I commented out a couple of SympyWriter test cases that combine the same symbol used differently. I do not understand the SympyWriter enough to fix it, when I tried it I end up with huge modifications to that class (maybe @arporter or @hiker could suggest if I am missing a simpler fix). But then the change is quite inconsequential because when this is used, e.g. for symbolic comparison, if the sympy fails, it evaluates to False anyway, so I don't think is worth adding a lot of extra complexity for this. So I would proposed deleting or postponing solving this case.

Reply:
> However, unlike the ArrayType, a RoutineSymbol has no information on the number of arguments

Btw, this PR only affects ArrayReferences that had generic Symbols or UnresolvedTypes, if it had an ArrayType and knew the shape, it is still an ArrayReference with the same type.

Reply:
I am not entirely certain what the expected behaviour should be. My understanding is that an unresolved expressions `c(i)` should be considered a call (and not an array access). From that point of view, `c + c(i)` is actually an invalid expression.

Or is the expectation that the writer should then try to 'reconsider' `c` to be an array, and converting it this way? To some degree, on a sympy level (where arrays get converted to function), this kind of shouldn't matter, but of course Fortran can have named parameter (`c(i=i)`), which I would guess will not work.

I agree that it just shouldn't crash (which should be easy todo by just catching the exception).

We can discuss tonight.

Reply:
OK, I found the root cause: the variable access information for `c(1)` returns an 'unknown' access, call.py line 323:
```
        if type(self.symbol) is Symbol:
            var_accesses.add_access(sig, AccessType.UNKNOWN, self.routine)
```
The result of this is that variable access info then does not detect the accesses to `c+c(1)` to include an array index (`1`), and as such the symbol for `c` is not specialised to become an array, and then doesn't get the `c(sympy!no_bounds, sympy_nobounds,1)` replacement done, which then causes a crash.

Reply:
As discussed, I think we *can* guarantee that `c + c(1)` does always imply that `c` is an array. The F2008 standard has:
<img width="970" height="259" alt="image" src="https://github.com/user-attachments/assets/208c4ee5-dfee-4667-b235-5b1873eea744" />
so the parentheses are optional on a call to a subroutine that has no arguments but they are not optional for a call to a function.
Could we extend the new frontend functionality to apply this logic?

Either way, I tried PSyAD with this branch with the kernel named in #2935 and it seemed to be fine (although that kernel may have changed a lot in the interim as I get nowhere near as many PSyclone-generated loop variables).

Reply:
I implemented the refinement and now "c + c(1)" is a "Reference + ArrayReference". However, the sympy evaluation still didn't like it be because it doesn't matter if it is an ArrayReference or a Call, both end up being SympyFunctions.

The real issue/fix was in sympy_writer.py::reference_node.

But since I already implemented the previously discussed Node type refinement, I also left it in this PR.